### DISCUSSION TITLE
MiV and printf to the Debugger Console

### DISCUSSION THREAD

**Discussion by rockoguido** [Mi-V-Soft-RISC-V/miv-rv32-documentation#5] [Bare Metal Embedded Software]
Working with a customer with a Bare Metal App on the Mi-V and there is no UART support.   Is there a way to map printfs to the Debugger Console in Soft Console?   

Looked around and no magic Macro jumped out at me..

Thx!
Dave

**Comment by nitindeshpande** [comment]
@rockoguido Do you mean that there is no UART IP used in the design and you want to transport the messages over debugger JTAG interface?

**Reply by rockoguido** [reply]
Hi. yes there is no UART IP in this design and we want to transport messages over JTAG interface, is this possible?

**Reply by JV73** [reply]
I'm looking into this too.  I've seen this done with a C++ application running under eclipse on linux, where STDOUT/STDIN where re-directed to an 'application console' within eclipse.  There are settings in the 'debug configuration' to allocate a console for STDOUT (which I have enabled) but I have not figured out how to add the 'application console' to the debug perspective.

**Reply by AntonKrug** [reply]
I feels to me that you want semihosting. But maybe you could be interested of doing printf through the debugger.  Where you do not need semihosting support, nor working uart. 

![image](https://user-images.githubusercontent.com/30338596/193325276-2392c10a-de4c-4c64-bcbd-0272e4073c20.png)

And in the printf you can reference variables etc...

![image](https://user-images.githubusercontent.com/30338596/193325568-a82f4ad1-b108-485b-9aee-1f0f7636750e.png)

**Reply by JV73** [reply]
Hm - interesting.  Would you need to apply this to every instance of printf in the code base ?  Where does the output end up ?  I tried this in SoftConsole 2021.3 with the 'first' printf in my program targeted to an RV32 implementation.  I don't see the string that should have been printed in either the 'build console' or the 'debugger console' or a 'popup'.

**Reply by JV73** [reply]
I tried enabling 'ARM Semihosting' in the debugger settings, but again, I don't see an 'application console' or messages in either the build or debug console tabs of the debug perspective.

**Reply by JV73** [reply]
I'm beginning to get the feeling the SoftConsole OpenOCD implementation does not have RISC-V Semihosting support implemented, and likewise, the current miv_riscv_hal/stdio library does not contain hooks for semihosting stdin/stdout/stderr via the debugger. This would be a great enhancement as it is very common for an embedded ARM to support a 'virtual console' through the debugger (JTAG) without implementing hardware (UART) to bring it out of the chip.

**Reply by AntonKrug** [reply]
Not really for every printf instance as this can be applied to a application which doesn't even have printf at all. But you have to apply this from debuger so sharing code between co-workers will lose your dynamic-printf 'breakpoints'. You would have to apply it everywhere where you want the debuger to do the its printf, independetly if the there is a printf in the code or not.

The output is shown in Debuger Console:

![image](https://user-images.githubusercontent.com/30338596/193702578-c69ebeb0-7c9e-410b-8ec9-05416cc4844c.png)

Not sure about current state of OpenOCD, but when the RISC-V semihosting was introduced to OpenOCD, then at that time I was experimenting with it (just for fun), the Mi-V HAL was not supporting it, so I tested RISC-V boards from different vendors and with their HALs, where the semihosting worked. But that was long time ago and now not sure if the semihosting still works or not.

**Comment by JV73** [comment]
I've got this working on an Avalanche Board with SoftConsole v2021.3  You can define you own 'output buffer' and use sprintf(buf, fmt, ...) or your own print function to write a zero terminated string into the buffer, and then call sys_writes() [see attached] to send it to the GDB console via OpenOCD running under SoftConsole.  This approach is 'relatively non-blocking' in that the action will auto-complete.  Reading input from the GDB console is accomplished using the sys_readc() [see attached] function to poll for input characters one at a time.  Currently, I find that this action will 'block' execution of the RISC-V code because the GDB console 'buffers' typed characters until the 'enter key' is pressed.  Then each call to sys_readc() will return one character until the 'enter' character is read.  Then a subsequent call to sys_readc() will again 'block' execution.   I'm looking for a way to tell OpenOCD to turn-off character buffering in the GDB console - anyone know the right command ? 
[RV32-SemiHosting.zip](https://github.com/Mi-V-Soft-RISC-V/miv-rv32-documentation/files/9761221/RV32-SemiHosting.zip)

Note: I added a global variable that defaults to 'disable' the semi-hosting console interface at power-up/boot so that the RISC-V program can 'free run' at boot time. I added an ISR that is triggered by a push-button to 'enable' the interface after a debugger is attached, and an 'exit' command to again disable the interface before closing the debugger. That way - the 'main loop' can operate normally when the console is not active.

Note: if you get 'strange' results when printing to the GDB console via semi-hosting, make sure that you're GDB console preferences look something like this: 
![GDB_preferences](https://user-images.githubusercontent.com/106761965/195248915-a77ae749-713f-4ede-81b4-9430427fbffc.JPG)
